### DISCUSSION TITLE
Perferred method for accessing database files as a Framework

### DISCUSSION THREAD

**Discussion by Trini8ed** [pocketbase/pocketbase#1888] [Q&A] [answered]
I'm using pocketbase as a framework and am wondering what the preferred way to access files that are in a collection.

Would the preferred route be to download the files through HTTP when using it as a framework or would it be better to just grab the files in pb_data/storage using the collection id and record id to load the files you need?

I won't be using S3 buckets but I assume if I did I would have to use the HTTP method?

![image](https://user-images.githubusercontent.com/1361448/220194570-68744c02-28ea-421f-bdc6-c4fd955fe3f6.png)

**Comment by ganigeorgiev** [comment, marked as answer]
You can access and read the file via the filesystem abstraction. Something like:

```go
fs, err := app.NewFilesystem()
if err != nil {
	return err
}
defer fs.Close()

record, err := app.Dao().FindRecordById("example", "ABC")
if err != nil {
	return err
}

fileKey := record.BaseFilesPath() + "/" + record.GetString("yourFileField")

br, err := fs.GetFile(fileKey)
if err != nil {
	return err
}
defer br.Close()
```

**Reply by Trini8ed** [reply]
Yup NewFilesystem was the call I was looking for thank you. I will go ahead and use that instead. Does NewFilesystem abstraction account for S3 buckets as well not just local file storage?

**Reply by ganigeorgiev** [reply]
>  Does NewFilesystem abstraction account for S3 buckets as well not just local file storage?

Yes, If you use S3 for storage then `fs.GetFile()` will make a S3 API call and will fetch the resource. The returned `*blob.Reader` implements `io.ReadSeekCloser` so you should be able to use it as "regular" file by reading/seeking its content.
